cwms-data-api/src/main/java/cwms/cda/data/dao/MeasurementDao.java

```diff
@@ -0,0 +1,500 @@
+
+import java.util.List;
+
+import static java.util.stream.Collectors.toList;
+import org.jooq.impl.DSL;
+import usace.cwms.db.dao.util.OracleTypeMap;
+import usace.cwms.db.jooq.codegen.packages.CWMS_STREAM_PACKAGE;
+import usace.cwms.db.jooq.codegen.udt.records.STREAMFLOW_MEAS2_T;
+import usace.cwms.db.jooq.codegen.udt.records.STREAMFLOW_MEAS2_TAB_T;
+
+public final class MeasurementDao extends JooqDao<Measurement> {
+    static final XmlMapper XML_MAPPER = buildXmlMapper();
+
+    public MeasurementDao(DSLContext dsl) {
+        super(dsl);
+    }
+
+    /**
+     * Retrieve a list of measurements
+     *
+     * @param officeId   - the office id
+     * @param locationId - the location id for filtering
+     * @param unitSystem - the unit system to use for the returned data
+     * @return a list of measurements
+     */
+    public List<Measurement> retrieveMeasurements(String officeId, String locationId, Instant minDateMask, Instant maxDateMask, String unitSystem) {
+        return connectionResult(dsl, conn -> {
+            setOffice(conn, officeId);
+            Number minHeight = null;
+            Number maxHeight = null;
+            Number minFlow = null;
+            Number maxFlow = null;
+            String minNum = null;
+            String maxNum = null;
+            String agencies = null;
+            String qualities = null;
+            Timestamp minTimestamp = OracleTypeMap.buildTimestamp(minDateMask == null ? null : Date.from(minDateMask));
+            Timestamp maxTimestamp = OracleTypeMap.buildTimestamp(maxDateMask == null ? null : Date.from(maxDateMask));
+            TimeZone timeZone = OracleTypeMap.GMT_TIME_ZONE;
+            STREAMFLOW_MEAS2_TAB_T retrieved = CWMS_STREAM_PACKAGE.call_RETRIEVE_MEAS_OBJS(DSL.using(conn).configuration(), locationId, unitSystem, minTimestamp, maxTimestamp,
+                    minHeight, maxHeight, minFlow, maxFlow, minNum, maxNum, agencies, qualities, timeZone.getID(), officeId);
+            return retrieved.stream()
+                    .map(MeasurementDao::fromJooqMeasurementRecord)
+                    .collect(toList());
+        });
+    }
+
+    /**
+     * Store a measurement
+     *
+     * @param measurement  - the measurement to store
+     * @param failIfExists - if true, fail if the measurement already exists
+     */
+    public void storeMeasurement(Measurement measurement, boolean failIfExists) {
+        connection(dsl, conn -> {
+            setOffice(conn, measurement.getOfficeId());
+            String failIfExistsStr = formatBool(failIfExists);
+            String xml = toDbXml(measurement);
+            CWMS_STREAM_PACKAGE.call_STORE_MEAS_XML(DSL.using(conn).configuration(), xml, failIfExistsStr);
+        });
+    }
+
+    /**
+     * Delete a measurement
+     *
+     * @param officeId   - the office id
+     * @param locationId - the location id of the measurement to delete
+     */
+    public void deleteMeasurements(String officeId, String locationId) {
```

Comment:
The null variables below should be passed in, except the Timestamps, those should be an Instant.  This provides more flexibility for the caller.  They can be added to the Controller later as optional parameters.

Reply:
Updated to parameterize the dao with null fields, so the controller will be responsible for deciding which fields are passed in/null.

Reply:
I disagree with the API choice here. We shouldn't be requiring null parameters for unused inputs. I'm asking Bryson to revert in:https://github.com/USACE/cwms-data-api/pull/890/files#diff-2e2e24ebae962a792534301fa3000a230f3241295a9110caa403c5f92e08264bR216.